### ISSUE TITLE
validate_presence_of fails for Shrine UploadedFile since empty string was introduced

### ISSUE THREAD

**Issue by klaustopher** [thoughtbot/shoulda-matchers#1228]
This somehow relates to #1223 and #1218, but I think it is also different. Please feel free to close and merge with one of the open issues.

Currently our update to `4.1.0` is blocked because on of our test cases is failing with the new version.

The model is super simple:

```ruby
class ImageUpload < ApplicationRecord
  include ImageUploader::Attachment.new(:image)
  validates_presence_of :image
end
```

I have played around a bit and the shrine uploader simply ignores assigning `""` to the image attribute:

```ruby
>> up = FactoryBot.build(:image_upload)
# => #<ImageUpload id: nil, image_data: {"id"=>"a4cd635bb1252991f6c539e3d27220cd.jpg", "storage"=>"cache", "metadata"=>{"filename"=>"image_upload_fixture.jpg", "size"=>12183, "mime_type"=>"image/jpeg", "width"=>100, "height"=>200}}, created_at: nil, updated_at: nil>

>> up.image
# => #<ImageUploader::UploadedFile:0x00007fb961af2bf8 @data={"id"=>"a4cd635bb1252991f6c539e3d27220cd.jpg", "storage"=>"cache", "metadata"=>{"filename"=>"image_upload_fixture.jpg", "size"=>12183, "mime_type"=>"image/jpeg", "width"=>100, "height"=>200}}>

>> up.image = ''
# => ""

>> up.image
# => #<ImageUploader::UploadedFile:0x00007fb961afaf88 @data={"id"=>"a4cd635bb1252991f6c539e3d27220cd.jpg", "storage"=>"cache", "metadata"=>{"filename"=>"image_upload_fixture.jpg", "size"=>12183, "mime_type"=>"image/jpeg", "width"=>100, "height"=>200}}>

>> up.image = nil
# => nil

>> up.image
# => nil
```

Checking all the conditions:

```ruby
>> ImageUpload.reflect_on_association(:image)
# => nil

>> ImageUpload.attribute_types['image']
# => #<ActiveModel::Type::Value:0x00007fd607362d58 @precision=nil, @scale=nil, @limit=nil>

>> ImageUpload.attribute_types['image'].respond_to?(:coder)
# => false
```

Shrine seems to use an unspecified `ActiveModel::Type` class.

---

For now, we have added a few custom tests instead of `validate_presence_of` for the field, but maybe it's possible to make the `empty string check` optional with a modifier?

**Comment by mcmire** [comment]
Thanks for the report! I have a branch in the works that _should_ fix this issue, but I'll leave this open in the meantime.

**Comment by vitalinfo** [comment]
the same problem with serializable fields, like
`serialize :data, Array`

got an exception:
```
     Failure/Error: it { is_expected.to validate_presence_of :data }
     
     ActiveRecord::SerializationTypeMismatch:
       can't serialize `data`: was supposed to be a Array, but was a String. -- ""
....
```

**Comment by benedikt** [comment]
I'm running into similar problems with attributes (or rather methods) that get validated but are not handling strings but objects. In theory I could make all setters handle empty strings, but it feels like the wrong path.

**Comment by mcmire** [comment]
@klaustopher @vitalinfo @benedikt Hey, I just pushed up a branch, `fix-presence-validation-for-activemodel`, that should fix this issue. Can you temporarily switch to it in your Gemfile and see if makes your tests pass?

**Comment by benedikt** [comment]
@mcmire Yes, that works. Thanks a lot for your efforts, Elliot!

**Comment by klaustopher** [comment]
Tested revision 11f4483 and it works fine with Shrine.

**Comment by mcmire** [comment]
Fixed in [v4.1.1](https://github.com/thoughtbot/shoulda-matchers/releases/tag/v4.1.1).
